pinot-broker/src/main/java/org/apache/pinot/broker/routing/instanceselector/StrictReplicaGroupInstanceSelector.java

```diff
@@ -0,0 +1,179 @@
+ * (Note that the replica-group information is derived from the ideal state of the table, where the instances are sorted
+ * alphabetically in the instance state map, so the replica-groups in the instance selector might not match the
+ * replica-groups in the instance partitions.) The instances in a replica-group should have all the online segments
+ * (segments with ONLINE/CONSUMING instances in the ideal state and selected by the pre-selector) available
+ * (ONLINE/CONSUMING in the external view) in order to serve queries. If any segment is unavailable in the
+ * replica-group, we mark the whole replica-group down and not serve queries with this replica-group.
+ *
+ * The selection algorithm is the same as {@link ReplicaGroupInstanceSelector}, and will always evenly distribute the
+ * traffic to all replica-groups that have all online segments available.
+ *
+ * The algorithm relies on the mirror segment assignment from replica-group segment assignment strategy. With mirror
+ * segment assignment, any server in one replica-group will always have a corresponding server in other replica-groups
+ * that have the same segments assigned. For example, if S1 is a server in replica-group 1, and it has mirror server
+ * S2 in replica-group 2 and S3 in replica-group 3. All segments assigned to S1 will also be assigned to S2 and S3. In
+ * stable scenario (external view matches ideal state), all segments assigned to S1 will have the same enabled instances
+ * of [S1, S2, S3] sorted (in alphabetical order). If we always pick the same index of enabled instances for all
+ * segments, only one of S1, S2, S3 will be picked, and all the segments are processed by the same server. In
+ * transitioning/error scenario (external view does not match ideal state), if a segment is down on S1, we mark all
+ * segments with the same assignment ([S1, S2, S3]) down on S1 to ensure that we always route the segments to the same
+ * replica-group.
+ * </pre>
+ */
+public class StrictReplicaGroupInstanceSelector extends ReplicaGroupInstanceSelector {
+
+  public StrictReplicaGroupInstanceSelector(String tableNameWithType, BrokerMetrics brokerMetrics) {
+    super(tableNameWithType, brokerMetrics);
+  }
+
+  /**
+   * {@inheritDoc}
+   *
+   * <pre>
+   * The maps are calculated in the following steps to meet the strict replica-group guarantee:
+   *   1. Create a map from online segment to set of instances hosting the segment based on the ideal state
+   *   2. Gather the online and offline instances for each online segment from the external view
+   *   3. Compare the instances from the ideal state and the external view and gather the unavailable instances for each
+   *      set of instances
+   *   4. Exclude the unavailable instances from the online instances map
+   * </pre>
+   */
+  @Override
+  void updateSegmentMaps(ExternalView externalView, IdealState idealState, Set<String> onlineSegments,
```

Comment:
high-level comment, since this is to calculate the map based on the coarse-grained unit (i.e. replica group) instead of the segment, have you considered mapping the online segments to the replica groups first, and then use the groups for selection?

Reply:
Good question. The reason why we do not pre-group the instances but use the same index to pick the instance for each segment is because the replica-group derived from the ideal state might not be the same as the replica-group from the instance assignment. If we pre-group the instances, when we shut down all instances from a replica-group, these instances might fall into multiple replica-groups derived from the ideal state, and causing multiple replica-groups down in the routing.

Reply:
I see. Does this mean the grouping is neither static nor deterministic? When you mentioned the mirroring above, I thought there is a pre-group concept.

Reply:
The grouping is not static, but is deterministic for each query. The grouping is determined by the index of the instance for each segment.

Reply:
Got it. Makes sense to me.